src/functions.php

```diff
@@ -12,22 +12,29 @@
  * Creates a `Promise` which resolves with the stream data buffer
  *
  * @param ReadableStreamInterface $stream
+ * @param int $maxLength
  * @return Promise\CancellablePromiseInterface Promise<string, Exception>
  */
-function buffer(ReadableStreamInterface $stream)
+function buffer(ReadableStreamInterface $stream, $maxLength = null)
 {
     // stream already ended => resolve with empty buffer
     if (!$stream->isReadable()) {
         return Promise\resolve('');
     }
 
+    $deferred = new Promise\Deferred();
     $buffer = '';
-    $bufferer = function ($data) use (&$buffer) {
+    $bufferer = function ($data) use (&$buffer, &$bufferer, $stream, $deferred, $maxLength) {
         $buffer .= $data;
+        if ($maxLength !== null && isset($buffer[$maxLength])) {
+            $deferred->reject(new \RuntimeException('Buffer exceeded maximum length'));
```

Comment:
This doesn't unregister the data callback, so it might result in a double resolution of the deferred, no?

Reply:
There is no double resolution in react/promise, first come, first serve. Subsequent resolutions are ignored.

There is room for improvement for this function in general, but that's something for a follow-up PR.

Reply:
What @jsor says, but also when rejecting there the error passes through https://github.com/WyriHaximus-labs/promise-stream/blob/7f505f3915e331b90b6e3a6040a33d4104243e15/src/functions.php#L49-L55 which does remove the listener